app/src/main/java/com/neo/speaktouch/utils/Reader.kt

```diff
@@ -22,159 +22,74 @@ import android.content.Context
             }
+        }
 
-            is Level.Children -> buildList {
-                for (child in node) {
-
-                    if (!NodeValidator.isValidForAccessibility(child)) continue
-                    if (!NodeValidator.isReadableAsChild(child)) continue
+        buildList {
+            add(content)
 
-                    add(
-                        getContent(
-                            child,
-                            Level.Text(
-                                mustReadSelection = false,
-                                mustReadType = listOf(
-                                    Type.SWITCH,
-                                    Type.TOGGLE,
-                                    Type.RADIO,
-                                    Type.CHECKBOX
-                                ).any { it == Type.get(child) }
-                            )
-                        )
-                    )
+            if (options.mustReadType) {
+                node.toText()?.let {
+                    add(it.resolved(context))
                 }
             }
-        }.joinToString(", ")
-    }
 
-    private fun getType(
-        node: AccessibilityNodeInfoCompat,
-        mustRead: Boolean
-    ): String? {
-
-        if (!mustRead) return null
-
-        return when (Type.get(node)) {
-            Type.NONE -> null
-            Type.IMAGE -> context.getString(R.string.text_image_type)
-            Type.SWITCH -> {
-                val switchType = context.getString(R.string.text_switch_type)
-                val selectionStatus = if (node.isChecked) {
-                    node.stateDescription?.toString() ?: context.getString(R.string.text_enabled)
-                } else {
-                node.stateDescription?.toString() ?: context.getString(R.string.text_disabled)
-            }
-                "$switchType, $selectionStatus"
-            }
-            Type.TOGGLE -> {
-                val toggleType = context.getString(R.string.text_toggle_type)
-                val selectionStatus = if (node.isChecked) {
-                    node.stateDescription?.toString() ?: context.getString(R.string.text_pressed)
-                } else {
-                node.stateDescription?.toString() ?: context.getString(R.string.text_not_pressed)
-            }
-                "$toggleType, $selectionStatus"
-            }
-            Type.RADIO -> {
-                val radioButtonType = context.getString(R.string.text_radio_type)
-                val selectionStatus = if (node.isChecked) {
-                    node.stateDescription?.toString() ?: context.getString(R.string.text_selected)
-                } else {
-                    node.stateDescription?.toString() ?: context.getString(R.string.text_not_selected)
-                }
-                "$radioButtonType, $selectionStatus"
+            if (options.mustReadSelection && node.isSelected) {
+                add(context.getString(R.string.text_selected))
```

Comment:
**doubt:** Can we also use state description in selection?

Reply:
From what I saw in the compose code, yes we can.

Reply:
@PatrykMis  We can extract all this logic involving stateDescription into a single function. What do you think?

Reply:
I think you're right, it will be cleaner and easier to maintain.

Reply:
- [ ] **to do:** extract state reading

Reply:
Talkback allows the combination of states, but I have decided for now to follow a different direction, as I understand that such combination makes the reading confusing. However, I have opened a discussion about this [^1].

[^1]: #115